### PR TITLE
[SPARK-13977] [SQL] Brings back Shuffled hash join

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

ShuffledHashJoin (also outer join) is removed in 1.6, in favor of SortMergeJoin, which is more robust and also fast.

ShuffledHashJoin is still useful in this case: 1) one table is much smaller than the other one, then cost to build a hash table on smaller table is smaller than sorting the larger table 2) any partition of the small table could fit in memory.

This PR brings back ShuffledHashJoin, basically revert #9645, and fix the conflict. Also merging outer join and left-semi join into the same class. This PR does not implement full outer join, because it's not implemented efficiently (requiring build hash table on both side).

A simple benchmark (one table is 5x smaller than other one) show that ShuffledHashJoin could be 2X faster than SortMergeJoin.
## How was this patch tested?

Added new unit tests for ShuffledHashJoin.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 1]
```diff
@@ -263,32 +263,20 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
       val df = df1.join(df2, $"key" === $"key2", "leftsemi")
       testSparkPlanMetrics(df, 1, Map(
-        0L -> ("LeftSemiJoinHash", Map(
+        0L -> ("ShuffledHashJoin", Map(
           "number of output rows" -> 2L)))
       )
     }
   }
 
-  test("LeftSemiJoinBNL metrics") {
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: LeftSemiJoinBNL is removed before, not sure why this test does not fail

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala [hunk 2]
```diff
@@ -247,7 +247,26 @@ class BenchmarkWholeStageCodegen extends SparkFunSuite {
...
+      val df2 = sqlContext.range(N / 5).selectExpr(s"id * 3 as k2")
+      df1.join(df2, col("k1") === col("k2")).count()
+    }
+
+    /**
+    Intel(R) Core(TM) i7-4558U CPU @ 2.80GHz
+    shuffle hash join:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+    -------------------------------------------------------------------------------------------
+    shuffle hash join codegen=false          1168 / 1902          3.6         278.6       1.0X
+    shuffle hash join codegen=true            850 / 1196          4.9         202.8       1.4X
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: SMJ will take about 450ns per row.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/LeftSemiJoinHash.scala [hunk 3]
```diff
@@ -1,58 +0,0 @@
...
-import org.apache.spark.sql.catalyst.plans.LeftSemi
-import org.apache.spark.sql.catalyst.plans.physical.{ClusteredDistribution, Distribution, Partitioning}
-import org.apache.spark.sql.execution.{BinaryNode, SparkPlan}
-import org.apache.spark.sql.execution.metric.SQLMetrics
-
-/**
- * Build the right table's join keys into a HashedRelation, and iteratively go through the left
- * table, to find if the join keys are in the HashedRelation.
- */
-case class LeftSemiJoinHash(
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/LeftSemiJoinHash.scala]: with the removal of this, can we now remove GeneralHashedRelation?

**Comment by rxin** [comment]
Can we add a config option to allow turning off shuffle hash join?

**Comment by davies** [comment]
@rxin Shuffle hash join should be turned using spark.sql.autoBroadcastThreshold, should we create another config for similar purpose?

**Comment by rxin** [comment]
I was thinking an option to prefer sort merge join. My main concern is that it is easier to oom when there are skew in hash joins.

**Comment by davies** [comment]
spark.sql.join.preferSortMergeJoin = true ?

**Comment by rxin** [comment]
sgtm

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/LeftSemiJoinHash.scala (hunk 3 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/LeftSemiJoinHash.scala]: I'd like to clean up HashedRelation as follow-up PR, ok ?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/LeftSemiJoinHash.scala]: sg

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 4]
```diff
@@ -101,10 +100,41 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
    *     [[org.apache.spark.sql.functions.broadcast()]] function to a DataFrame), then that side
    *     of the join will be broadcasted and the other side will be streamed, with no shuffling
    *     performed. If both sides of the join are eligible to be broadcasted then the
+   * - Shuffle hash join: if single partition is small enough to build a hash table.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: if the average size of a single partition - you can change that in your next pr though if this is the only minor comment

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoin.scala [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Expression, JoinedRow}
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.physical._
+import org.apache.spark.sql.execution.{BinaryNode, SparkPlan}
+import org.apache.spark.sql.execution.metric.SQLMetrics
+
+/**
+ * Performs an inner hash join of two child relations by first shuffling the data using the join
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoin.scala]: inner?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 6]
```diff
@@ -134,6 +176,18 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
         Seq(joins.BroadcastHashJoin(
           leftKeys, rightKeys, RightOuter, BuildLeft, condition, planLater(left), planLater(right)))
 
+      case ExtractEquiJoinKeys(LeftOuter, leftKeys, rightKeys, condition, left, right)
+         if !conf.preferSortMergeJoin && canBuildHashMap(right) && muchSmaller(right, left) ||
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: can't we use shouldShuffleHashJoin here?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: ah ok the problem is you need an extra if

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 7]
```diff
@@ -118,6 +148,18 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           leftKeys, rightKeys, Inner, BuildLeft, condition, planLater(left), planLater(right)))
 
       case ExtractEquiJoinKeys(Inner, leftKeys, rightKeys, condition, left, right)
+        if !conf.preferSortMergeJoin && shouldShuffleHashJoin(left, right) ||
+          !RowOrdering.isOrderable(leftKeys) =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: this is not part of your pr -- but looking at this reminds me that we should make sure this is never true (!isOderable) -- keys that cannot be sorted should just not be join keys.

**Comment by rxin** [comment]
lgtm - please address the minor issues in ur next pr.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 7 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: join key and grouping key should not require having order, right?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: no - but all the keys that can be hashed i think can be sorted too.

**Comment by davies** [comment]
Merging this into master, those minor comments will be addressed in follow-up PRs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 8]
```diff
@@ -101,10 +100,41 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
+   * - Shuffle hash join: if single partition is small enough to build a hash table.
    * - Sort merge: if the matching join keys are sortable.
    */
   object EquiJoinSelection extends Strategy with PredicateHelper {
 
+    /**
+     * Matches a plan whose single partition should be small enough to build a hash table.
+     */
+    def canBuildHashMap(plan: LogicalPlan): Boolean = {
+      plan.statistics.sizeInBytes < conf.autoBroadcastJoinThreshold * conf.numShufflePartitions
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: `conf.numShufflePartitions` only works if the number of shuffle partitions is a constant value for a query, right? Can you add a comment at here?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Good point, will add a comment.

**Pull request closed without merging**